src/com/google/cloud/datastore/types.js

```diff
@@ -229,6 +229,82 @@ foam.LIB({
               JSON.stringify(v));
         };
       })()
+    },
+    {
+      name: 'toDatastoreKeyName',
+      code: foam.mmethod({
+        Number: function(n) { return n.toString(); },
+        String: function(str) { return str; },
+        FObject: function(o) {
+          var idProp = o.cls_.ID;
+          if ( ! idProp ) {
+            throw new Error('Attempt to construct datastore key from ' +
+                'unidentified object');
+          }
+          return idProp.toDatastoreKeyName(o);
+        },
+        Array: function(a) {
+          throw new Error('Multi-part keys must be derived from objects, not values');
+        },
+        Object: function(o) {
+          throw new Error('Cannot convert plain object to datastore key name');
```

Comment:
I just realized, we should probably special case `Date` objects here; provide an implementation consistent with `foam.core.Date.toDatastoreKeyNamePart(o)` below.